### ISSUE TITLE
val detect

### ISSUE THREAD

**Issue by Ellohiye** [ultralytics/ultralytics#22532]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Are the example files obtained by executing val.py in yolov8 and the detection results of detect.py on the same image?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Ellohiye, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will review your issue and assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including:
-Exact commands used for validation and detection (CLI or Python)
-Model and version (e.g., YOLO11n/YOLO11s), checkpoints, and any custom weights
-Sample input image(s) and the expected vs. observed outputs
-Validation/predict configs (confidence, NMS IoU, image size, augment/test-time settings)
-Environment details (OS, Python, PyTorch, CUDA) and logs

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
There are no `val.py` or `detect.py` executable script in Ultralytics. That's from YOLOv5 repository.

**Comment by Ellohiye** [comment]
I made some modifications, which are basically just the two commands: yolo detect and yolo val.

**Comment by glenn-jocher** [comment]
Thanks—no, not by default. `yolo detect val` evaluates on the `val` split from your `data.yaml`, while `yolo detect predict` runs on whatever `source` you provide; they’ll be the same only if you point `predict` to the same validation images.

```bash
# metrics on the validation split
yolo detect val model=path/to/best.pt data=path/to/data.yaml

# run inference on the exact val images (use the 'val' path from your data.yaml)
yolo detect predict model=path/to/best.pt source=/path/to/val/images
```

See the Val mode guide for details in [Model Validation with Ultralytics YOLO](https://docs.ultralytics.com/modes/val/) and API specifics in the [DetectionValidator reference](https://docs.ultralytics.com/reference/models/yolo/detect/val/).

**Comment by Ellohiye** [comment]
So, val_batch0_pred.jpg and the same image get the same result when detected using detect, right? Because both val and detect use the same parameters for the image.

**Comment by glenn-jocher** [comment]
Not necessarily—the defaults differ: `val` uses `conf=0.001` while `predict` defaults to `conf=0.25`; match your `conf/iou/imgsz` to mirror `val` (and note `val_batch0_pred.jpg` is a batch collage, not a single image). See predict defaults in the [Model Prediction guide](https://docs.ultralytics.com/modes/predict/) and validation defaults in the [Configuration – Validation Settings](https://docs.ultralytics.com/usage/cfg/). ([docs.ultralytics.com](https://docs.ultralytics.com/modes/predict/))

```bash
# Mirror val-style thresholds on a single image
yolo detect predict model=path/to/best.pt source=/path/to/that_image.jpg conf=0.001 iou=0.7 imgsz=640
```

**Comment by Ellohiye** [comment]
However, testing revealed that the image val_batch0_pred.jpg did not contain any detection boxes with a confidence level less than 0.25, and the accuracy of the detected target was completely consistent with that of the detected target.

**Comment by glenn-jocher** [comment]
Makes sense—this can happen: `val_batch*_pred.jpg` is drawn from NMS-filtered outputs, so if all surviving boxes are >0.25 they’ll look identical to `predict` at 0.25; to verify 1:1, save labels from both modes with identical thresholds and diff the TXT files.

```bash
# save per-image predictions from val
yolo detect val model=path/to/best.pt data=path/to/data.yaml imgsz=640 conf=0.001 iou=0.7 save_txt=True project=runs/compare name=val

# run predict on the same val images and save labels
yolo detect predict model=path/to/best.pt source=/path/to/val/images imgsz=640 conf=0.001 iou=0.7 save_txt=True project=runs/compare name=predict
```

Details: NMS and thresholds are applied in [postprocess()](https://docs.ultralytics.com/reference/models/yolo/detect/val/#postprocess), the collage is drawn in [plot_predictions()](https://docs.ultralytics.com/reference/models/yolo/detect/val/#plot_predictions), and per-image labels are written via [save_one_txt()](https://docs.ultralytics.com/reference/models/yolo/detect/val/#save_one_txt).

**Comment by Ellohiye** [comment]
Thank you for your detailed answer. I would like to ask why batch detection using val and single-image detection using predict might lead to different results?

**Comment by glenn-jocher** [comment]
Good question—results can differ because `val` runs batched, rectangular letterboxing and may use the model’s remembered `imgsz` and FP16, while single-image `predict` letterboxes that one image (often to a slightly different shape) and defaults to FP32; those small resize/precision changes can shift scores and NMS outcomes. To make them match, align preprocessing and thresholds:

```bash
# Force identical preprocessing (square letterbox) and precision on both
yolo detect val model=path/to/best.pt data=path/to/data.yaml imgsz=640 rect=False half=False conf=0.25 iou=0.7 save_txt=True project=runs/align name=val
yolo detect predict model=path/to/best.pt source=/path/to/val/images imgsz=640 rect=False half=False conf=0.25 iou=0.7 save_txt=True project=runs/align name=predict
```

**Comment by Ellohiye** [comment]
If these parameters are consistent and val uses the parameter bs=1 for inference, will the resulting accuracy be completely consistent?

**Comment by glenn-jocher** [comment]
Yes—if you use `batch=1` (not `bs`) and perfectly match `imgsz/rect/half/conf/iou/max_det/classes/agnostic_nms` on the same images, `val` and `predict` should yield identical detections up to tiny floating‑point differences; `val` runs a fixed, non‑shuffled loader and the same preprocessing pipeline, see [get_dataloader()](https://docs.ultralytics.com/reference/models/yolo/detect/val/#get_dataloader) and [build_dataset()](https://docs.ultralytics.com/reference/models/yolo/detect/val/#build_dataset); if you still see drift, please share the exact commands and the two saved label folders for comparison.

**Comment by Ellohiye** [comment]
Is the pred image output by detectval really not based on a confidence threshold of 0.25? I see the confusion matrix is ​​output based on a threshold of 0.25.

**Comment by glenn-jocher** [comment]
Not 0.25—`yolo detect val` draws `val_batch*_pred.jpg` from NMS outputs using your `conf` setting (default `conf=0.001`), and the confusion matrix uses the same `conf`; if you’re seeing 0.25 there, it came from the run’s arguments or an earlier run—rerun with your desired `conf` and check the new outputs.

**Comment by Ellohiye** [comment]
I started a new thread, and an official replied saying, "The Confusion matrix is ​​generated at conf=0.25 and with multi-label prediction enabled. Default prediction doesn't use multi-label."

**Comment by glenn-jocher** [comment]
Thanks for the follow-up—clarifying: in `yolo detect val` the confusion matrix uses the run’s `conf` value (not hardcoded to 0.25), and `val` runs NMS with `multi_label=True`; you can see the CM threshold passed as `conf=self.args.conf` in [DetectionValidator.update_metrics](https://docs.ultralytics.com/reference/models/yolo/detect/val/#update_metrics). If you want CM at 0.25, rerun with `conf=0.25` (e.g., `plots=True` to regenerate figures).
